addons/web_editor/tests/test_controller.py

```diff
@@ -0,0 +1,61 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+import binascii
+
+from odoo.addons.http_routing.models.ir_http import slug
+import odoo.tests
+from odoo.tests.common import HttpCase
+
+@odoo.tests.tagged('-at_install', 'post_install')
+class TestController(HttpCase):
+    def test_01_illustration_shape(self):
+        self.authenticate('admin', 'admin')
+        # SVG with all replaceable colors.
+        svg = b"""
+<svg viewBox="0 0 400 400">
+  <rect width="300" height="300" style="fill:#3AADAA;" />
+  <rect x="20" y="20" width="300" height="300" style="fill:#7C6576;" />
+  <rect x="40" y="40" width="300" height="300" style="fill:#F6F6F6;" />
+  <rect x="60" y="60" width="300" height="300" style="fill:#FFFFFF;" />
+  <rect x="80" y="80" width="300" height="300" style="fill:#383E45;" />
+</svg>
+        """
+        # Need to bypass security check to write image with mimetype image/svg+xml
+        context = {'binary_field_real_user': self.env['res.users'].sudo().browse([1])}
+        attachment = self.env['ir.attachment'].sudo().with_context(context).create({
+            'name': 'test.svg',
+            'mimetype': 'image/svg+xml',
+            'datas': binascii.b2a_base64(svg, newline=False),
+            'public': True,
+            'res_model': 'ir.ui.view',
+            'res_id': 0,
+        })
+        # Shape illustration with slug.
+        url = '/web_editor/shape/illustration/%s' % slug(attachment)
+        palette = 'c1=%233AADAA&c2=%237C6576&&c3=%23F6F6F6&&c4=%23FFFFFF&&c5=%23383E45'
+        attachment['url'] = '%s?%s' % (url, palette)
+
+        response = self.url_open(url)
+        self.assertEqual(200, response.status_code, 'Expect response')
+        self.assertEqual(svg, response.content, 'Expect unchanged SVG')
+
+        response = self.url_open(url + '?c1=%23ABCDEF')
+        self.assertEqual(200, response.status_code, 'Expect response')
+        self.assertEqual(len(svg), len(response.content), 'Expect same length as original')
+        self.assertTrue('ABCDEF' in str(response.content), 'Expect patched c1')
+        self.assertTrue('3AADAA' not in str(response.content), 'Old c1 should not be there anymore')
+
+        # Shape illustration without slug.
+        url = '/web_editor/shape/illustration/noslug'
+        attachment['url'] = url
+
+        response = self.url_open(url)
+        self.assertEqual(200, response.status_code, 'Expect response')
+        self.assertEqual(svg, response.content, 'Expect unchanged SVG')
+
+        response = self.url_open(url + '?c1=%23ABCDEF')
+        self.assertEqual(200, response.status_code, 'Expect response')
+        self.assertEqual(len(svg), len(response.content), 'Expect same length as original')
+        self.assertTrue('ABCDEF' in str(response.content), 'Expect patched c1')
+        self.assertTrue('3AADAA' not in str(response.content), 'Old c1 should not be there anymore')
```

Comment:
@bso-odoo I have something mysterious about this test, could you explain? Without your fix, the test rightfully fails... but not at the first assert of this code, but the one at line 60.

What I have at the url_open at line 53 (without your fix):
- Goes through the shape controller -> raise not found :+1: 
- the response of line 53 is a 200 response :shrug: :shrug: :shrug: :shrug: :shrug:

Reply:
Maybe this is related to the strange fallback mechanism in the (pre apo) http layer ???
I have to replace
```py
raise werkzeug.exceptions.NotFound()
```
by
```py
return request.not_found()
```
otherwise it generates the response from the `create()` that happened within the same session (because of the way the test environment is made).

There are 100 `raise*NotFound` in the code which might have to be replaced if their route is able to fail after having created something.

Reply:
It was ultimately related to the `_serve_fallback()` mechanism, which ends up using `get_serve_attachment()` which finds the attachment based on its URL.

Reply:
> EDIT: ah I see you changed the "raise not found" into "return not found" anyway (so one go through server_fallback and not the other?). I don't know which one is the best grimacing @rdeodoo ?

I guess it is better to re-enable the http-level fallback: what if another mechanism would match the request for other reasons, just like `get_server_attachment()` does.

Reply:
If I got this right, from the commit messages from httpocalypse, old codebase and @JKE-be quick discussion, the return vs raise is just about going into the fallback stack and so:
1. Using or not the fallback mechanism (attachment, redirect..)
2. Responding quicker or not

There seems to be no good or bad use, both serves a purpose.
In the code base there is 50/50 use of return vs raise and I think most of them were not thought about but just written like that.

Maybe @Julien00859 will have some interesting input here :+1:

Reply:
> The other difference is that the "revived" domain relies on an "=like" while the http-level fallback uses an "=" url.
It would feel safer (and faster) to use "=": the request path can typically contain '%' characters for other reasons.

That was definitely a bug in the original implementation, I'm opening a PR to fix this in 14.0

Reply:
https://github.com/odoo/odoo/pull/86623

Reply:
Since httpocalypse (1) `raise werkzeug.exceptions.NotFound` is preferred over `return request.not_found` inside of controllers, (2) the fallback mechanism only applies to `_match()` (and its override), (3) a `_match()` override still can bypass the fallback by returning `NotFound()` instead of raising it (not used at the moment).

In master the fallback mechanism was applied for all 404-not-found exceptions, not matter where the error was thrown.

> I had in mind that we did not fallback on that if we matched a controller first

Is true since httpocalypse.

Related code is:
* (v15/before httpocalypse) `_handle_exception` in `base/ir.http`
* (master/after httpocalypse) `_serve_ir_http` in `core/http`